### ISSUE TITLE
How is Angular2 bundled

### ISSUE THREAD

**Issue by Ludo09** [angular/angular#7228]
Dear Angular team,

Could you please explain how you create Angular2.js bundle?

I'm wondering because each .ts file contains no indication of the name of the bundle it belongs to.

For instance we write `import from 'angular2/http'` however no http/*.ts contain any information that it belongs to this module.

Do you produce any `d.ts` that can be used later on by an external Typescript program?

I'm also wondering what is the pipeline / workflow & toolset which is used to create `angular2.js`?

**Comment by aciccarello** [comment]
Does this help?

https://github.com/angular/angular/blob/master/modules/angular2/docs/bundles/overview.md

**Comment by Ludo09** [comment]
Thanks but unfortunately it doesn't really help. I read this overview before I posted my questions. What's unclear is what are the chain of steps & tools that used to build the different bundles, and in which order. 

I have to say that as usual with tools like Gulp and Grunt (especially when scripts become that big), it becomes very hard to follow.

**Comment by aciccarello** [comment]
Yeah, I'm not sure if there is a whole explanation of the gulp file. I'd probably try to decode the [build task](https://github.com/angular/angular/blob/master/gulpfile.js#L1409) but is is hard to follow everything.

**Comment by pkozlowski-opensource** [comment]
@Ludo09 we are using https://github.com/systemjs/builder to create bundles in the System.register format and WebPack to create UMD ones.

Bundles are build from entry points that are located, respectively:
* for System.register ones: https://github.com/angular/angular/blob/e7470d557df0400a812d84551524970cb5f12eee/gulpfile.js#L144-L160
* for UMD ones:  https://github.com/angular/angular/tree/master/tools/build/webpack

As you've noticed our bundling setup is not trivial and I don't think I can offer more guidance without more concrete questions from your end. All I can say is that you need to be pretty familiar with both SystemJS builder and WebPack to make sense of those build config and scripts.

Going to close this one as we are not planning on documenting build steps in more details at this point. Feel free to ask more concrete questions though.

**Comment by Ludo09** [comment]
@pkozlowski-opensource I went through the Gulp build script and figured out that Webpack config was actually part of it as well (no external webpack.conf.js).

Could you please explain:
- why there is a tsconfig.json where `module: "commonjs"`
- at the same time webpack config contains tsc config section where `module: "commonjs"`
- at the same time webpack config also contains webpack config where `libraryTarget: "umd"`?
Which one is used to do what? 
Does it work like this: tsc compiles .ts in es5/commonjs and then webpack bundle everything into es5/umd?

**Comment by pkozlowski-opensource** [comment]
> Does it work like this: tsc compiles .ts in es5/commonjs and then webpack bundle everything into es5/umd?

Yes, we can simplify it like this. WebPack is used to combine ES5 output. At the point of bundling TypeScript is not involved at all.

**Comment by antonybudianto** [comment]
@Ludo09 you could take a look at my starter project [here](github.com/antonybudianto/angular2-starter), it uses SystemJS Builder, and the gulp tasks is split to files so it's easy to learn

**Comment by Ludo09** [comment]
@pkozlowski-opensource thanks. Why exactly does Angular2 use Webpack and SystemJS?

@antonybudianto thanks but 
1. I'm getting 404 - page not found (correct url is [https://github.com/antonybudianto/angular2-starter/](https://github.com/antonybudianto/angular2-starter)) 
2. actually if you would have the same for webpack and npm only (no Gulp) I'd be very interested ;-)

**Comment by SirKnightDragoon** [comment]
@Ludo09 @antonybudianto For the bundle it's more this one:
https://github.com/antonybudianto/angular-webpack-starter
